src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/NumericIndexWriter.java

```diff
@@ -0,0 +1,178 @@
+
+        this.indexDescriptor = indexDescriptor;
+        this.indexContext = indexContext;
+        this.bytesPerValue = bytesPerValue;
+        this.writer = new BlockBalancedTreeWriter(maxSegmentRowId + 1, bytesPerValue, maxPointsInLeafNode);
+    }
+
+    @Override
+    public String toString()
+    {
+        return MoreObjects.toStringHelper(this)
+                          .add("bytesPerDim", bytesPerValue)
+                          .add("bufferedPoints", writer.getPointCount())
+                          .toString();
+    }
+
+    public static class LeafCallback implements BlockBalancedTreeWriter.Callback
+    {
+        final List<PackedLongValues> postings = new ArrayList<>();
+
+        public int numLeaves()
+        {
+            return postings.size();
+        }
+
+        @Override
+        public void writeLeafDocs(BlockBalancedTreeWriter.RowIDAndIndex[] sortedByRowID, int offset, int count)
+        {
+            PackedLongValues.Builder builder = PackedLongValues.monotonicBuilder(PackedInts.COMPACT);
+
+            for (int i = offset; i < count; ++i)
+            {
+                builder.add(sortedByRowID[i].rowID);
+            }
+            postings.add(builder.build());
+        }
+    }
+
+    /**
+     * Writes a balanced tree and posting lists from a {@link org.apache.lucene.codecs.MutablePointValues}.
+     *
+     * @param values points to write
+     *
+     * @return metadata describing the location and size of this balanced tree in the overall SSTable balanced tree component file
+     */
+    public SegmentMetadata.ComponentMetadataMap writeAll(IntersectingPointValues values) throws IOException
+    {
+        long treePosition;
+        SegmentMetadata.ComponentMetadataMap components = new SegmentMetadata.ComponentMetadataMap();
+
+        LeafCallback leafCallback = new LeafCallback();
+
+        try (IndexOutput treeOutput = indexDescriptor.openPerIndexOutput(IndexComponent.BALANCED_TREE, indexContext, true))
+        {
+            // The SSTable balanced tree component file is opened in append mode, so our offset is the current file pointer.
+            long treeOffset = treeOutput.getFilePointer();
+
+            treePosition = writer.writeField(treeOutput, values, leafCallback);
+
+            // If the treePosition is less than 0 then we didn't write any values out
+            // and the index is empty
+            if (treePosition < 0)
+                return components;
+
+            long treeLength = treeOutput.getFilePointer() - treeOffset;
+
+            Map<String, String> attributes = new LinkedHashMap<>();
+            attributes.put("max_points_in_leaf_node", Integer.toString(writer.getMaxPointsInLeafNode()));
+            attributes.put("num_leaves", Integer.toString(leafCallback.numLeaves()));
+            attributes.put("num_points", Long.toString(writer.getPointCount()));
+            attributes.put("bytes_per_dim", Long.toString(writer.getBytesPerValue()));
```

Comment:
Could this be named "bytes_per_value", like the getter, having only one dimension?